### ISSUE TITLE
Tune name confidence and hamming distance thresholds

### ISSUE THREAD

**Issue by LittleRichard** [LittleRichard/sormtger#49]
Tune the parameters using real images:
- StaticCardDAO
  - MIN_DIST_THRESHOLD
  - HIGH_CONFIDENCE_THRESHOLD
  - NEAR_HIGH_CONFIDENCE_THRESHOLD
- SetHashDAO
  - DIST_MATCH_THRESHOLD
  - HIGH_CONFIDENCE_DIST
  - NEAR_HIGH_CONFIDENCE_DIST

**Comment by LittleRichard** [comment]
@Panzer23 , started testing against your images and the results were decent but not great without any extra effort.  I have plenty to demo, and i think we have some options here, so let me know when you can be free to wander over!

**Comment by LittleRichard** [comment]
I poked around a bit more last night and thought about the causes, and I think there are 3 issues:
* The old cards' font is giving the OCR a tough time
  * I'll do more trial runs against the less-ornate modern font tonight, it's likely that it will be way better but I don't think this is 100% of the solution.
  * I'm looking into training all three fonts as a language to give to the OCR
* The pictures are too dark, so the resolution seems to have been downgraded to capture low-light pictures.  OCR seems to be having trouble getting clean edges, and will need higher contrast (more light) and less blurriness (ISO 100 with enough light)
  * I'm optimistic that better light and some camera tricks will make a big difference.
* The alignment tolerance isn't good enough for the symbols.  Symbol location crops were not picking out consistent areas of the card.  I still need to tune the crop selection a bit, but that won't fix the issue.

**Comment by Panzer23** [comment]
@LittleRichard 

1. Card font,  there are a couple places we can find the fonts online, I think I might have sent you an email with some links way back.  Ill see if I can find it.

2. fair enough,  I  don't think the focus is "perfect" either.  need to do a bit of tweaking,  Ill set up a bigger light with a diffuser to deal with reflections  #45 will help with the light as well.

3.  thats interesting.  im not sure what i can do about that, might need a rethink of the pic-tray if that is the case not sure what  I can do to make it any more consistent in this current configuration.

**Comment by Panzer23** [comment]
@LittleRichard I have wednesday blocked off to do robot things with eric (s saldano bf).  come on over if you want that day.

**Comment by LittleRichard** [comment]
I can't do anything after work Wednesday, so maybe once you're back from abroad.

Card font - already found all 3 and am working on a training set

Camera stuff - all good things, don't get too weird until you try more light, I really think it will work wonders.  Plus we can tweak some of the camera settings in python too

Alignment - i'll get some more demo images before passing a complete judgment, but it wasn't very close in the few i tried

**Comment by Panzer23** [comment]
@LittleRichard Ok I will see you after I get back then.  
The light is something I will get on, i prolly just need to put it in a light box and get it over with.  Pretty easy to do. If I do that I can make conditions consistent too, all the better for tuning the camera. 

Fair enough about the alignment, we knew this was gonna be the sticking point, we will get it working.  would it be better if the decision tree were something like:

Read name > check possible sets > use box setting from those sets? 
Read name > check set > if possible OG frame, check frame > If OG, use cropbox 1, else use cropbox 2?

dunno if that's overly complicated, but it might be a way to pare down the potential problems.

**Comment by LittleRichard** [comment]
Light box would rule, but just more of it will be enough to confirm image improvement.  It should be pretty obvious that it's helping if you compare a well-lit tight zoom on the text to some of the stuff i emailed you in the tarball.  A set of 10-20 images with better light would be extremely helpful here.  Also, would you mind adjusting the camera a tiiiiiny bit, there's a consistent skew of a couple degrees that you can see in the name crops

Looks like plenty of room for camera tricks afterward, I was able to squeeze a 10% accuracy boost out of the name (from 50% to 60%) just by increasing the contrast in python.  But let's keep image manipulation out of the picamera settings, we may want to do contrast for just the name, i'm not sure.

Agreed on the decision tree, it's the most direct solution but will involve a lot of hard-coding box settings; it will sort of suck to code and maintain as new sets are added, and isn't very clean performance-wise.  I think you're right though, it's probably the safest way.

**Comment by LittleRichard** [comment]
Some more useful stuff.

https://pypi.python.org/pypi/tesserocr (see 'Advanced API Examples')
* The tesseract python wrapper library we use has some de-skewing stuff built in
* It also lets you define the PSM 'page segmentation mode', which defaults to full page of text (eek!)

https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality#image-processing

**Comment by Panzer23** [comment]
sweet stuff!  the PSM sounds like it should help define the area a bit more.  Also sounds like a starbucks drink.

I think we should be ok with the skewing, or at least be able to minimize it, I will put that second on the list of priorities. Once we get the prototype working, I can create an overlay within picamera that is square, and then adjust the camera for each unit with the overlay on, so each one is aligned quite well.

**Comment by LittleRichard** [comment]
Merged the following changes to master.  Symbol stuff is still very broken, and name is better, though wasn't working well against the current test set; will re-visit with better-lit and less-skewed images once they exist.  Even without direct evidence of progress, lots of learning here around tesseract innards.

All the hard-mode tesseract optimizations that I could find.
* whitelisting characters for tesseract to [a-zA-z0-9] and punctuation [,.:;/'?()&-]
* turned off the 'learning' of a tesseract API object over time
  * apparently it gets wonky sometimes after it's seen a ton of data, and we'll be using a singleton instance of the API over the server's lifetime
* turned off dictionary-match confidence score bonus/demerits, makes sense since many words we'll match aren't english.  Though many will be (goblin, cleric), the ones that aren't (rakdos, purphoros) will be more important to get right to differentiate between different types of goblins/clerics
* added page-segmentation-mode setting to the singleton tesseract API, tested and found the most performant
  * bonus, it comes with skew detection/handling without sacrificing too much performance (as near as i could tell)

Added some image pre-processing stuff for the name:
* contrast increase
* grayscaling.  Could do earlier in the process, but i think it's good to do it just-in-time because:
  * non-name crops may not want grayscale (they do, but w/e)
  * wasteful to grayscale the full image and then cropping 90% of it off
    * even if done multiple times on the same image for 2+ rotations, still likely to be more efficient.

Tuned the tesseract word confidence parameter threshold, we can re-visit but i think it's good enough.

Tuned the postgres fuzzy-match threshold, we can re-visit but i think it's good enough.

Added and tuned (for v.05 camera placement) the crop boxes for 0/180 for name/symbol.  We'll have to fix these with any hardware changes, but not a big deal; the important part was to get them defined as variables and we can deal with the values later.  Also added the placeholders for 90/270 crop boxes, but we'll handle those "You//Choose" cards later.

Increased the symbol hash bit count from 8 to 16.  2^16 should give more range of values, 2^8 wasn't really enough to cover all the different symbols with respectable hamming distances between them all.  

Added a bunch of debug logging as well as saving cropped images to disk for examination against the logs

**Comment by LittleRichard** [comment]
Closing for now, will re-open for next round of tuning

**Comment by LittleRichard** [comment]
While looking at pyocr in #53 , found an interesting library that promises to enhance text a bit:
https://github.com/jflesch/libpillowfight#stroke-width-transformation

**Comment by LittleRichard** [comment]
Will need to re-visit with the change to openCV
